crates/CrateAssembler.kt

```diff
@@ -0,0 +1,213 @@
+        }
+
+        outputMetadataFile.outputStream().use {
+            val metadata = constructMetadata()
+            it.write(metadata.toString().toByteArray(StandardCharsets.UTF_8))
+        }
+    }
+
+    private fun constructMetadata(): JsonObject {
+        return JsonObject().apply {
+            set("name", name)
+            set("vers", versionFile.readText())
+            val depsArray = JsonArray()
+            for (dep in deps) {
+                val (depName, depVer) = dep.split("=")
+                val obj = JsonObject()
+                obj.set("optional", false)
+                obj.set("default_features", false)
+                obj.set("name", depName)
+                obj.set("features", JsonArray())
+                obj.set("version_req", depVer)
+                obj.set("target", Json.NULL)
+                obj.set("kind", "normal")
+                obj.set("registry", "https://github.com/rust-lang/crates.io-index")
+                depsArray.add(obj)
+            }
+            set("deps", depsArray)
+            set("features", JsonObject())
+            set("authors", JsonArray().apply { authors.filter { it != "" }.forEach { add(it) } })
+            set("description", description)
+            set("documentation", documentation)
+            set("homepage", homepage)
+            readmeFile?.let {
+                set("readme", readmeFile?.readText())
+                set("readme_file", it.toPath().fileName.toString())
+            } ?: run {
+                set("readme", Json.NULL)
+                set("readme_file", Json.NULL)
+            }
+            set("keywords", JsonArray().apply { keywords.filter { it != "" }.forEach { add(it) } })
+            set("categories", JsonArray().apply { categories.filter { it != "" }.forEach { add(it) } })
+            set("license", license)
+            set("license_file", Json.NULL)
+            set("repository", repository)
+            // https://doc.rust-lang.org/cargo/reference/manifest.html#the-badges-section
+            // as docs state all badges should go to README so it's safe to keep it empty
+            set("badges", JsonObject())
+            set("links", Json.NULL)
+        }
+    }
+
+    private fun generateCargoToml(): String {
+        val mapper = TomlMapper()
+        return mapper.writeValueAsString(
+            CargoToml(
+                CargoTomlPackage(
+                    name,
+                    edition,
+                    versionFile.readText(),
+                    authors.filter { it != "" },
+                    homepage,
+                    repository,
+                    documentation,
+                    description,
+                    readmeFile?.toPath()?.fileName?.toString() ?: ""
+                ),
+                deps.associate { it.split("=").let { (dep, ver) -> dep to ver } }
+            )
+        )
+    }
+}
+
+fun main(args: Array<String>): Unit = exitProcess(CommandLine(CrateAssembler()).execute(*args))
```

Comment:
Let's add a newline

Reply:
@lolski addressed in https://github.com/vaticle/bazel-distribution/pull/322/commits/df4f472de180f956b9a57ee7688d49b038382153